src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyProgressService.java

```diff
@@ -217,26 +204,187 @@ public CompetencyProgress updateCompetencyProgress(Long competencyId, User user)
     }
 
     /**
-     * Calculate the progress value for the given user in a competency.
+     * Calculate the progress for the given user in a competency.
+     * The progress for exercises is the percentage of points achieved in all exercises linked to the competency.
+     * The progress for lecture units is the percentage of lecture units completed by the user.
+     * The final progress is a weighted average of the progress in exercises and lecture units.
+     *
+     * @param lectureUnits                  The lecture units linked to the competency
+     * @param exerciseInfos                 The information about the exercises linked to the competency
+     * @param numberOfCompletedLectureUnits The number of lecture units completed by the user
+     * @param competencyProgress            The progress entity to update
+     */
+    private void calculateProgress(Set<LectureUnit> lectureUnits, Set<CompetencyExerciseMasteryCalculationDTO> exerciseInfos, int numberOfCompletedLectureUnits,
+            CompetencyProgress competencyProgress) {
+        double numberOfLearningObjects = lectureUnits.size() + exerciseInfos.size();
+        if (numberOfLearningObjects == 0) {
+            // If nothing is linked to the competency, the competency is considered completed
+            competencyProgress.setProgress(100.0);
+        }
+
+        double achievedPoints = exerciseInfos.stream().mapToDouble(info -> info.lastPoints() != null ? info.lastPoints() : 0).sum();
+        double maxPoints = exerciseInfos.stream().mapToDouble(CompetencyExerciseMasteryCalculationDTO::maxPoints).sum();
+        double exerciseProgress = maxPoints > 0 ? achievedPoints / maxPoints * 100 : 0;
+
+        double lectureProgress = 100.0 * numberOfCompletedLectureUnits / lectureUnits.size();
+
+        double progress = exerciseInfos.size() / numberOfLearningObjects * exerciseProgress + lectureUnits.size() / numberOfLearningObjects * lectureProgress;
```

Comment:
Splitting it up into 2 variables increases the readability for me

```suggestion
        double exerciseProgress = exerciseInfos.size() / numberOfLearningObjects * exerciseProgress;
        double lectureProgress = lectureUnits.size() / numberOfLearningObjects * lectureProgress;
        double progress = exerciseProgress + lectureProgress;
```